presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -462,6 +506,48 @@ private int compare(Type type, Object first, Object second)
         }
     }
 
+    /**
+     * Replace time zone component of a {@link TimestampWithTimeZoneType} value with a given one.
+     */
+    private static Object withTimeZone(TimestampWithTimeZoneType type, Object value, TimeZoneKey newZone)
+    {
+        if (type.isShort()) {
+            return packDateTimeWithZone(unpackMillisUtc((long) value), newZone);
+        }
+        LongTimestampWithTimeZone longTimestampWithTimeZone = (LongTimestampWithTimeZone) value;
+        return LongTimestampWithTimeZone.fromEpochMillisAndFraction(longTimestampWithTimeZone.getEpochMillis(), longTimestampWithTimeZone.getPicosOfMilli(), newZone);
+    }
+
+    private static TimeZoneKey getTimeZone(TimestampWithTimeZoneType type, Object value)
+    {
+        if (type.isShort()) {
+            return unpackZoneKey(((long) value));
+        }
+        return TimeZoneKey.getTimeZoneKey(((LongTimestampWithTimeZone) value).getTimeZoneKey());
+    }
+
+    @VisibleForTesting
+    static boolean isTimestampToTimestampWithTimeZoneInjectiveAt(ZoneId zone, Instant instant)
+    {
+        ZoneOffsetTransition transition = zone.getRules().previousTransition(instant);
+        if (transition != null && !transition.getDuration().isNegative() &&
+                transition.getDateTimeAfter().minusNanos(1).atZone(zone).toInstant().isAfter(instant)) {
+            // DST change forward and the instant is ambiguous, being within the 'gap' area non-monotoic remapping
+            return false;
+        }
+        return true;
+    }
+
+    private static Instant getInstantTruncation(TimestampWithTimeZoneType type, Object value)
```

Comment:
What does "instant truncation" refer to here?

Reply:
sorry, this is a typo. Changed to `getInstantWithTruncation`

(the point is -- there is no rounding involved when going from picos to nanos)